parquet/src/main/java/org/apache/iceberg/parquet/ReusableMapData.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.parquet;
+
+import java.util.Map;
+
+public interface ReusableMapData {
+  ReusableArrayData keys();
+  ReusableArrayData values();
+
+  default void grow() {
+    keys().grow();
+    values().grow();
+  }
+
+  default int capacity() {
+    return keys().capacity();
+  }
+
+  default void setNumElements(int numElements) {
+    keys().setNumElements(numElements);
+    values().setNumElements(numElements);
+  }
+
+  int size();
+
+  @SuppressWarnings("unchecked")
+  default <K, V> Map.Entry<K, V> getRaw(
+      int pos,
+      ParquetValueReaders.ReusableEntry<K, V> reuse,
+      ParquetValueReaders.ReusableEntry<K, V> def) {
+    if (pos < capacity()) {
+      reuse.set((K) keys().getObj(pos), (V) values().getObj(pos));
+      return reuse;
+    }
+    return def;
+  }
+
+  default void addPair(Object key, Object value) {
+    if (size() >= capacity()) {
+      grow();
+    }
+    keys().update(size(), key);
+    values().update(size(), value);
+    setNumElements(size() + 1);
```

Comment:
Does this need to set this every time a new pair is added? I think it would be better to set the valid size just before returning, to avoid extra work.

Reply:
I might be misinterpreting your question / concern, but it seems like at least in the `SparkReusableMapData` this is the case. 

If I'm reading this correctly, addPair is overwritten and then `setNumElements` is only called on `buildMap`. Again, I might be misinterpreting the question / concern but for a default implementation given that it seems to be overridden on the Spark side at least.

```java
private static class MapReader<K, V> extends RepeatedKeyValueReader<MapData, SparkReusableMapData, K, V> {
    // Removed to the bare minimum ....
    @Override
    protected void addPair(SparkReusableMapData map, K key, V value) {
      if (writePos >= map.capacity()) {
        map.grow();
      }

      map.keys.values[writePos] = key;
      map.values.values[writePos] = value;

      writePos += 1;
    }

    @Override
    protected MapData buildMap(SparkReusableMapData map) {
      map.setNumElements(writePos);
      return map;
    }
```

Reply:
Yes, my point is that we don't need to set the number of valid elements until the end, just before MapData is returned.

Reply:
So yes we could do that, but we'd be keeping track of the write position in the `MapReader` so this just goves the grow code into the base class. I can swap it back around though.

Reply:
Oh right because this is an interface we need to call a method anyways since we can't have a concrete type in the interface.